### PR TITLE
[CBRD-23173] Loaddb must use BU_LOCK

### PR DESCRIPTION
**Pull request by lucianlazar2211**
http://jira.cubrid.org/browse/CBRD-23173

After the introduction of the new `BU_LOCK` we had to use it for the new loaddb. Now workers will acquire `BU_LOCK` on the class at the beginning of their transaction. Also, we disable class locking during `heap_insert_logical` and we make sure that the worker has acquired `BU_LOCK` beforehand.

### CODE REVIEW

File: src/storage/heap_file.c [hunk 1]
```diff
@@ -22355,9 +22355,17 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
    */
   /* make sure we have IX_LOCK on class see [NOTE-1] */
```
Comment (eseokoh) [inline src/storage/heap_file.c]: please move this inside of else block.

File: src/storage/heap_file.c [hunk 2]
```diff
@@ -22355,9 +22355,17 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
...
   // todo - this should probably be disabled in case of bulk insert
```
Comment (eseokoh) [inline src/storage/heap_file.c]: please remove obsoleted.

File: src/storage/heap_file.c [hunk 3]
```diff
@@ -22355,9 +22355,17 @@ heap_insert_logical (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context,
...
-  if (lock_object (thread_p, &context->class_oid, oid_Root_class_oid, IX_LOCK, LK_UNCOND_LOCK) != LK_GRANTED)
+  if (context->is_bulk_op)
     {
-      return ER_FAILED;
+      // In case of bulk insert we need to skip the IX lock on class and make sure that we have BU_LOCK acquired.
+      assert (locator_is_BU_locked (context->class_oid));
```
Comment (ribram) [inline src/storage/heap_file.c]: replace locator_is_BU_locked with lock_has_lock_on_object.  locator_is_BU_locked was a work-around used to trick system into thinking that loaddb workers have BU locks :). Now that they really do, it shouldn't be needed.

File: src/transaction/locator_sr.c [hunk 4]
```diff
@@ -13723,11 +13723,24 @@ xlocator_demote_class_lock (THREAD_ENTRY * thread_p, const OID * class_oid, LOCK
   return lock_demote_class_lock (thread_p, class_oid, lock, ex_lock);
 }
 
-static bool
+bool
 locator_is_BU_locked (const OID & class_oid)
```
Comment (eseokoh) [inline src/transaction/locator_sr.c]: I think we can replace it with `lock_has_lock_on_object` that is a general function for the purpose.

File: src/communication/network_interface_sr.c [hunk 5]
```diff
@@ -84,6 +84,7 @@
 #include "xasl_cache.h"
 #include "elo.h"
 #include "transaction_transient.hpp"
+#include "network_interface_sr.h"
```
Comment (eseokoh) [inline src/communication/network_interface_sr.c]: Do you still need it?

Comment (lucianlazar2211) [reply src/communication/network_interface_sr.c]: Nope. I am removing it. :)

**Pull request merged**